eip-0014.md

```diff
@@ -510,3 +510,239 @@ PoolNFT        | Coll[Byte] | Pool NFT ID
     sigmaProp(Pk || (validPoolIn && validReturnOut))
 }
 ```
+
+### Ergo AMM DEX Contracts [Ergo to Token]
+
+The Ergo-to-token exchange is essentially the native-to-token (N2T) contract of UniSwap 1.0.
+There are two approaches to create a N2T exchange:
+1. Use a T2T (token-to-token) exchange, where one of the tokens maps to Ergs and have a separate dApp that exchanges Ergs to tokens at 1:1 rate.
+2. Implement N2T directly in the exchange contract. Here we use this approach.
+
+The N2T AMM DEX relies on two types of contracts as before:
+
+- Pool contracts
+- Swap contracts
+
+#### Pool contracts
+
+The following is the modified pool contract representing a Liquidity Pool of the N2T AMM DEX.
+As before, the pool contract ensures the following operations are performed according to protocol rules:
+
+- Depositing. An amount of LP tokens taken from LP reserves is proportional to an amount of underlying assets deposited. `LP = min(X_deposited * LP_supply / X_reserved, Y_deposited * LP_supply / Y_reserved)`
+- Redemption. Amounts of underlying assets redeemed are proportional to an amount of LP tokens returned. `X_redeemed = LP_returned * X_reserved / LP_supply`, `Y_redeemed = LP_returned * Y_reserved / LP_supply`
+- Swap. Tokens are exchanged at a price corresponding to a relation of a pair’s reserve balances while preserving constant product constraint (`CP = X_reserved * Y_reserved`). Correct amount of protocol fees is paid (0.03% currently). `X_output = X_reserved * Y_input * 997 / (Y_reserved * 1000 + Y_input * 997)`
+
+Variables:
+- `X_deposited` - Amount of the first asset (nanoErgs) being deposited to the pool box
+- `Y_deposited` - Amount of the second asset being deposited to the pool box
+- `X_reserved` - Amount of the first asset (nanoErgs) locked in the pool box
+- `Y_reserved` - Amount of the second asset locked in the pool box
+- `LP_supply` - LP tokens circulating supply corresponding to the pool box
+
+#### Schema of the pool UTXO
+
+Section   | Description
+----------|------------------------------------------------------
+value     | Asset X reserves (nanoErgs)
+tokens[0] | Pool NFT
+tokens[1] | Locked LP tokens
+tokens[2] | Asset Y reserves
+R4[Long]  | Fee multiplier numerator (e.g. 0.003% fee -> 997 fee_num). This represents the *non-fee* part of the sold asset
+
+#### Pool contract
+
+```scala
+{
+    val InitiallyLockedLP = 0x7fffffffffffffffL
+    val FeeDenom = 1000
+    val minStorageRent = 10000000L  // this many number of nanoErgs are going to be permanently locked
```

Comment:
Why permanently locked? Why LP can't redeem all the liquidity and destroy the pool? Otherwise unless we consider this locked reserves in deposit / redeem formulas there will be some portion of LP tokens that can never be returned.

Reply:
That can never happen even in the T2T contracts. The moment you have successor condition, the contract will fail if there is no successor, and so we will just let that small amount of LP stuck till miner can claim it